crates/cheatcodes/tests/random.rs

```diff
@@ -0,0 +1,164 @@
+            .unwrap();
+        let end = out.find("Suite result: ok.").unwrap();
+        &out[start..end]
+    }
+
+    // Run the test twice with the same seed and verify the outputs are the same.
+    let seed1 = "0xa1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2";
+    let out1 = cmd
+        .args(["test", "--fuzz-seed", seed1, "-vv"])
+        .assert_success()
+        .get_output()
+        .stdout_lossy();
+    let res1 = extract_test_result(&out1);
+
+    let out2 = cmd
+        .forge_fuse()
+        .args(["test", "--fuzz-seed", seed1, "-vv"])
+        .assert_success()
+        .get_output()
+        .stdout_lossy();
+    let res2 = extract_test_result(&out2);
+
+    assert_eq!(res1, res2);
+
+    // Run the test with another seed and verify the output differs.
+    let seed2 = "0xb1b2c3d4e5f6a7b8c9d0e1f2a3b4c5d6e7f8a9b0c1d2e3f4a5b6c7d8e9f0a1b2";
+    let out3 = cmd
+        .forge_fuse()
+        .args(["test", "--fuzz-seed", seed2, "-vv"])
+        .assert_success()
+        .get_output()
+        .stdout_lossy();
+    let res3 = extract_test_result(&out3);
+    assert_ne!(res3, res1);
+
+    // Run the test without a seed and verify the outputs differs once again.
+    cmd.forge_fuse();
+    let out4 = cmd.args(["test", "-vv"]).assert_success().get_output().stdout_lossy();
+    let res4 = extract_test_result(&out4);
+    assert_ne!(res4, res1);
+    assert_ne!(res4, res3);
+});
+
+// tests `freshStorage` cheatcode.
+forgetest_init!(test_fresh_storage, |prj, cmd| {
+    prj.wipe_contracts();
+    prj.insert_ds_test();
+    prj.insert_vm();
+    prj.clear();
+
+    prj.add_source(
+        "Counter.t.sol",
+        r#"pragma solidity 0.8.24;
+import {Vm} from "./Vm.sol";
+import {DSTest} from "./test.sol";
+contract Counter {
+    uint256 public a;
+    address public b;
+    int8 public c;
+    address[] public owners;
+
+    function setA(uint256 _a) public {
+        a = _a;
+    }
+
+    function setB(address _b) public {
+        b = _b;
+    }
+
+    function getOwner(uint256 pos) public view returns (address) {
+        return owners[pos];
+    }
+
+    function setOwner(uint256 pos, address owner) public {
+        owners[pos] = owner;
+    }
+}
+
+contract CounterFreshStorageTest is DSTest {
+    Vm vm = Vm(HEVM_ADDRESS);
+
+    function test_fresh_storage() public {
+        uint256 index = 55;
+        Counter counter = new Counter();
+        vm.freshStorage(address(counter));
+
+        // Next call would fail with array out of bounds without freshStorage
+        address owner = counter.getOwner(index);
+        // Subsequent calls should retrieve same value
+        assertEq(counter.getOwner(index), owner);
+        // Change slot and make sure new value retrieved
+        counter.setOwner(index, address(111));
+        assertEq(counter.getOwner(index), address(111));
+
+        require(counter.a() != 0);
+        require(counter.b() != address(0));
+        require(counter.c() != 0);
```

Comment:
I actually don't see why these should hold.... As far as I can see, `a`, `b`, and `c` should all be fresh, meaning they _could_ be zero as well as any other number.

Reply:
yes, that's correct, they could be 0 though is unlikely. I will remove these assertions as they could fail the test

Reply:
@ehildenb I removed the assertions and added a test using seed / producing deterministic values here:
https://github.com/foundry-rs/foundry/blob/46d23223687ea373500ac852e892645eefe2d997/crates/cheatcodes/tests/random.rs#L198-L216

Also ported Kontrol tests using same seed:
https://github.com/foundry-rs/foundry/blob/46d23223687ea373500ac852e892645eefe2d997/crates/cheatcodes/tests/random.rs#L246-L277

Reply:
Nice, that's cool.